File: _architecture/reference/client_flutter/lib/core/logging/governor_logger.dart
// client_flutter — Governor Logger (Phase 12)
// Structured binary HBP log emitter over HTTP POST to the Rust Governor.
//
// Architecture Decision:
//   The spec originally called for sending log frames over the Socket.io connection
//   via SduiSocketManager. However, SduiSocketManager uses the socket_io_client package
//   which wraps all data in Socket.io protocol framing — raw binary HBP frames cannot
//   be injected without corrupting the SDUI data channel. The SDUI socket is also
//   module-scoped (it connects to /ws/shua_diary), while logs go to the Governor.
//
//   Solution: Fire-and-forget HTTP POST to $governorBaseUrl/api/logs/ingest.
//   Body = 12-byte HBP header + MessagePack-encoded integer-keyed log map.
//   This keeps the log channel cleanly decoupled from the SDUI data path.
//
// Zero-overhead fast-path:
//   Gate check (O(1) int compare) before ANY allocation.
//   If level < _minLogLevel, returns immediately — no Map allocation, no HTTP call.
//
// Singleton pattern: GovernorLogger() always returns the same instance.

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:msgpack_dart/msgpack_dart.dart' as msgpack;
import 'package:client_flutter/core/network/hbp_constants.g.dart';
import 'package:client_flutter/sdui/core/sdui_socket_provider.dart'
    show kGovernorBaseUrl;

// ─────────────────────────────────────────────────────────────────────────────
// HBP frame constants
// ─────────────────────────────────────────────────────────────────────────────

const int _kHbpMagicH = 0x48; // 'H'
const int _kHbpMagicB = 0x42; // 'B'
const int _kHbpTypeLog = 0x12; // Type: LOG

// ─────────────────────────────────────────────────────────────────────────────
// GovernorLogger — Singleton
// ─────────────────────────────────────────────────────────────────────────────

/// Structured HBP binary log emitter for the Flutter client.
///
/// Sends log frames to `POST $kGovernorBaseUrl/api/logs/ingest`.
/// Fire-and-forget: never awaits the HTTP response, never throws to the caller.
///
/// Usage:
/// ```dart
/// GovernorLogger().log(
///   HbpLogLevel.INFO,
///   'sdui_screen',
///   'Blueprint assembled in ${elapsed}ms',
///   tags: HbpLogTag.SDUI | HbpLogTag.PERF,
///   telemetry: {'elapsed_ms': elapsed},
/// );
/// ```
class GovernorLogger {
  // Dart's factory constructor + static field implements the singleton pattern
  // with zero overhead after first construction — subsequent calls return the
  // same memory address, no allocation.
  static final GovernorLogger _instance = GovernorLogger._internal();
  factory GovernorLogger() => _instance;
  GovernorLogger._internal();

  // Minimum level gate — configurable at runtime if needed.
  // Defaults to DEBUG (2) in debug mode for verbosity, otherwise INFO (3).
  int _minLogLevel = kDebugMode ? HbpLogLevel.DEBUG : HbpLogLevel.INFO;

  /// Update the client-side minimum log level threshold.
  void setMinLogLevel(int level) => _minLogLevel = level;

  // Pre-built ingest URI — computed once at first log call, not per-call.
  Uri? _ingestUri;
  Uri get _uri {
    _ingestUri ??= Uri.parse('$kGovernorBaseUrl/api/logs/ingest');
    return _ingestUri!;
  }

  /// Emits a structured binary log entry to the Governor.
  ///
  /// [level]      - HbpLogLevel integer (use HbpLogLevel.INFO, .WARN, etc.)
  /// [subsystem]  - Sub-component label (e.g. 'sdui_renderer', 'governor_client')
  /// [msg]        - Human-readable description of the event
  /// [tags]       - Optional 24-bit bitmask of HbpLogTag values
  /// [customTags] - Optional ad-hoc string tag array
  /// [telemetry]  - Optional key-value performance data map
  /// [traceId]    - Optional 8-char hex correlation ID
  void log(
    int level,
    String subsystem,
    String msg, {
    int tags = 0,
    List<String>? customTags,
    Map<String, dynamic>? telemetry,
    String? traceId,
  }) {
    // ── 1. Emitter Gate Check (O(1) — zero allocation fast path) ─────────────
    if (level < _minLogLevel) return;

    // ── 2. Build integer-keyed payload Map ────────────────────────────────────
    // Integer keys match BorrowedLogEntry serde(rename) mappings in Rust.
    final Map<int, dynamic> payload = {
      0: DateTime.now().millisecondsSinceEpoch, // ts
      1: level, // level
      2: HbpModule.CLIENT_FLUTTER, // module (11)
      3: subsystem, // subsystem
      4: msg, // msg
      5: tags, // tags
    };

    if (customTags != null) payload[6] = customTags;
    if (telemetry != null) payload[7] = telemetry;
    if (traceId != null) payload[8] = traceId;

    // ── 3. Encode + build HBP frame + POST (fire-and-forget) ─────────────────
    _sendAsync(payload);
  }

  /// Builds the 12-byte HBP header + MsgPack payload and POSTs it asynchronously.
  /// Any errors are silently dropped — the logger must never crash the app.
  void _sendAsync(Map<int, dynamic> payload) {
    // Dart's unawaited async — schedules on the event loop, never blocks the caller.
    () async {
      try {
        final Uint8List msgpackBytes = msgpack.serialize(payload);

        // Build 12-byte header:
        //   [0]    = 0x48 ('H') magic
        //   [1]    = 0x42 ('B') magic
        //   [2]    = 0x00       version (reserved)
        //   [3]    = 0x12       type: LOG
        //   [4-7]  = 0x00000000 flags / stream ID (reserved)
        //   [8-11] = payload length as u32 big-endian
        final header = ByteData(12);
        header.setUint8(0, _kHbpMagicH);
        header.setUint8(1, _kHbpMagicB);
        header.setUint8(3, _kHbpTypeLog);
        header.setUint32(8, msgpackBytes.length, Endian.big);

        final frame = Uint8List(12 + msgpackBytes.length);
        frame.setRange(0, 12, header.buffer.asUint8List());
        frame.setRange(12, 12 + msgpackBytes.length, msgpackBytes);

        await http
            .post(
              _uri,
              headers: const {'Content-Type': 'application/octet-stream'},
              body: frame,
            )
            .timeout(const Duration(seconds: 2));
      } catch (_) {
        // Silently drop — network errors, timeout, Governor offline:
        // none of these should ever surface to the application layer.
      }
    }();
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Convenience Extension (optional ergonomic API)
// ─────────────────────────────────────────────────────────────────────────────

/// Top-level singleton accessor for terseness in call sites.
/// `gLog.info('sdui_screen', 'Rendered in ${elapsed}ms', tags: HbpLogTag.PERF);`
final GovernorLogger gLog = GovernorLogger();

File: shua_modules/shua_diary/src/lib/governor_logger.ts
// shua_diary — Governor Logger v3.0
// Binary HBP structured log emitter over TCP loopback to shua_governor port 5001.
// Ported from horAIzon 2.0 governor_logger.ts — SDUI tag removed.
//
// Zero-overhead fast-path: level gate check BEFORE any allocation.
// If level < MIN_LOG_LEVEL, returns immediately with zero heap pressure.

import { encode } from '@msgpack/msgpack';
import * as net from 'net';

// ─────────────────────────────────────────────────────────────────────────────
// HBP Log Level constants (mirrors LEVEL_* in entry.rs and HbpLogLevel in Dart)
// ─────────────────────────────────────────────────────────────────────────────
export const HBP_LOG_LEVEL = Object.freeze({
  TRACE:    1,
  DEBUG:    2,
  INFO:     3,
  WARN:     4,
  ERROR:    5,
  CRITICAL: 6,
} as const);

// 24-bit bitmask log tag constants (mirrors HbpLogTag in hbp_constants.g.dart)
export const HBP_LOG_TAG = Object.freeze({
  SYSTEM:    1,
  LIFECYCLE: 2,
  NETWORK:   4,
  DATABASE:  8,
  PERF:      32,
  SECURITY:  64,
  AI:        128,
} as const);

// Module ID for shua_diary (mirrors HBP_MODULE in Rust/Dart)
const MODULE_ID_SHUA_DIARY = 3; // shua.diary module ID

// ─────────────────────────────────────────────────────────────────────────────
// TCP Loopback Socket to Governor log listener (port 5001)
// ─────────────────────────────────────────────────────────────────────────────
const LOG_SOCKET_PORT = 5001;

let socket: net.Socket | null = null;
let isConnecting = false;
const pendingQueue: Buffer[] = [];
const MAX_PENDING_QUEUE = 500;

function connectSocket(): void {
  if (socket || isConnecting) return;
  isConnecting = true;

  const s = new net.Socket();
  s.unref(); // Allow event loop to exit even with open socket
  s.setKeepAlive(true, 5000);

  s.once('connect', () => {
    socket = s;
    isConnecting = false;
    // Flush queued logs from bootstrap phase
    while (pendingQueue.length > 0) {
      const frame = pendingQueue.shift();
      if (frame) s.write(frame);
    }
  });

  const onError = (_err: Error) => {
    socket = null;
    isConnecting = false;
    setTimeout(connectSocket, 5000).unref(); // Retry after 5s
  };

  s.on('error', onError);
  s.once('close', () => {
    socket = null;
    isConnecting = false;
    setTimeout(connectSocket, 5000).unref();
  });

  s.connect(LOG_SOCKET_PORT, '127.0.0.1');
}

// Initiate connection immediately on module load
connectSocket();

// ─────────────────────────────────────────────────────────────────────────────
// Emitter Gate
// ─────────────────────────────────────────────────────────────────────────────
const MIN_LOG_LEVEL: number = parseInt(process.env.LOG_MIN_LEVEL ?? '3', 10);

// HBP frame magic bytes
const HBP_MAGIC_H  = 0x48; // 'H'
const HBP_MAGIC_B  = 0x42; // 'B'
const HBP_TYPE_LOG = 0x12; // Type: LOG = 18

// ─────────────────────────────────────────────────────────────────────────────
// Core log() function
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Emits a structured binary HBP log frame to the Governor socket.
 *
 * @param level      - HBP_LOG_LEVEL integer (1=TRACE … 6=CRITICAL)
 * @param subsystem  - Sub-component label (e.g. 'diary_repository', 'cert_repository')
 * @param msg        - Human-readable event description
 * @param tags       - Optional 24-bit bitmask of HBP_LOG_TAG values (default 0)
 * @param telemetry  - Optional key-value performance data
 * @param traceId    - Optional 8-char hex correlation ID for cross-module tracing
 */
export function log(
  level: number,
  subsystem: string,
  msg: string,
  tags: number = 0,
  telemetry?: Record<string, unknown>,
  traceId?: string,
): void {
  // ── 1. Emitter Gate (O(1) integer compare — zero allocation fast path) ──
  if (level < MIN_LOG_LEVEL) return;

  // ── 2. Build integer-keyed payload ────────────────────────────────────
  const payload: Record<string, unknown> = {
    '0': Date.now(),            // ts:        u64 Unix ms
    '1': level,                 // level:     u8
    '2': MODULE_ID_SHUA_DIARY,  // module:    u8  = 3 (shua.diary)
    '3': subsystem,             // subsystem: string
    '4': msg,                   // msg:       string
    '5': tags,                  // tags:      u32 bitmask
  };

  if (telemetry !== undefined) payload['7'] = telemetry;
  if (traceId   !== undefined) payload['8'] = traceId;

  // ── 3. Encode + write to socket ───────────────────────────────────────
  try {
    const msgpackBytes = encode(payload);

    // 12-byte HBP header
    const header = Buffer.alloc(12, 0);
    header.writeUInt8(HBP_MAGIC_H, 0);
    header.writeUInt8(HBP_MAGIC_B, 1);
    header.writeUInt8(HBP_TYPE_LOG, 3);
    header.writeUInt32BE(msgpackBytes.length, 8);

    const frame = Buffer.concat([header, msgpackBytes]);

    if (socket && !socket.destroyed) {
      socket.write(frame, (err) => {
        if (err && pendingQueue.length < MAX_PENDING_QUEUE) {
          pendingQueue.push(frame);
        }
      });
    } else {
      if (pendingQueue.length < MAX_PENDING_QUEUE) {
        pendingQueue.push(frame);
      }
    }
  } catch {
    // Logger must NEVER throw or crash the application thread.
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Convenience wrappers
// ─────────────────────────────────────────────────────────────────────────────

export const logger = {
  trace: (sub: string, msg: string, opts?: { tags?: number; telemetry?: Record<string, unknown>; traceId?: string }) =>
    log(HBP_LOG_LEVEL.TRACE, sub, msg, opts?.tags, opts?.telemetry, opts?.traceId),
  debug: (sub: string, msg: string, opts?: { tags?: number; telemetry?: Record<string, unknown>; traceId?: string }) =>
    log(HBP_LOG_LEVEL.DEBUG, sub, msg, opts?.tags, opts?.telemetry, opts?.traceId),
  info: (sub: string, msg: string, opts?: { tags?: number; telemetry?: Record<string, unknown>; traceId?: string }) =>
    log(HBP_LOG_LEVEL.INFO, sub, msg, opts?.tags, opts?.telemetry, opts?.traceId),
  warn: (sub: string, msg: string, opts?: { tags?: number; telemetry?: Record<string, unknown>; traceId?: string }) =>
    log(HBP_LOG_LEVEL.WARN, sub, msg, opts?.tags, opts?.telemetry, opts?.traceId),
  error: (sub: string, msg: string, opts?: { tags?: number; telemetry?: Record<string, unknown>; traceId?: string }) =>
    log(HBP_LOG_LEVEL.ERROR, sub, msg, opts?.tags, opts?.telemetry, opts?.traceId),
  critical: (sub: string, msg: string, opts?: { tags?: number; telemetry?: Record<string, unknown>; traceId?: string }) =>
    log(HBP_LOG_LEVEL.CRITICAL, sub, msg, opts?.tags, opts?.telemetry, opts?.traceId),
};
